### DISCUSSION TITLE
How do I make Bevy alterate a 2D Sprite material/lighting in complex ways?

### DISCUSSION THREAD

**Discussion by deavid** [bevyengine/bevy#6898] [Q&A]
I'm trying to make a game in isometric view with 2D assets painted in watercolor, and the game needs to be unsettling (terror, etc), so dynamic lighting is very important here.

I need to understand how to change how a sprite can be rendered, what is possible, what tools do I actually have.

So far I only encountered Alpha blending (both from PNG and overall transparency), and color tint. I think I saw masking as well, but unsure.

These fall a bit short for what I want; it is doable but I don't really like the path I would need to take.

I do have already a custom ray-tracing alike algorithm that is able to define the brightness per tile; and I could even add for direction of the light too. And colors.

Let me add here the things that I am looking for, and let me know if any can be done and where to look to learn about them.

* Advanced blending modes: Screen, Multiply, Overlay and such (Similar on how Photoshop/Gimp/Krita work for layer blending)
* Masked tints or lighting effects: I have a tile, and either is lit or unlit. It would be nice to see techniques to do gradients or other higher resolution approaches. All I need so far is to define the 4 corners with 4 values and a gradient should be applied as a masking for the effect or regulating the strength of it.
* Parsing normal maps for 2D sprites: would be nice if I could have a secondary image with vivid colors defining roughly where the pixel normal is facing, such that I can use that to compute light/darkening spots of the image. 
* Z-Buffer modifiers for 2D sprites: Some of the sprites represent very different parts of near and far; sometimes it is hard to resolve. If I could add a z-buffer to the sprite that adds a slight bump to the front or to the back it would help a lot blending more complex sprites. Otherwise, separating sprites into smaller ones is tedious, and joining them in screen might look weird with seams being hard to hide.

Most of the effects I'm looking for are for pre-processing the data of the sprite itself before rendering, then rendered in one shot, not to be applied in layers in screen. We could say that I'm looking for some sort of "imagemagik/gimp" to work on the fly for the sprite just before it is rendered. Also, most of those effects can be cached as new images if needed; as they're going to change slowly. (or even I could precompute a huge collection of options to grab from there; I might need maybe 20 variations of the same sprite only)

The main problem right now is that sprite tint only darkens it, what would benefit me the most right now would be a different tint that works like "screen" or "overlay", so I don't need 2 sprites to represent "normal lighting" plus "very bright lighting".

I think that what I'm asking are shader-related but I never used anything on that. If you can recommend me any exercise to try, it would be nice too. But well, the problem is that I have no idea what I'm asking, what I should be searching for any of this.

**Comment by wilk10** [comment]
I agree that Sprites won't cover well your usecase.

I see a couple of paths you could take:

1) you say you could precomute everything in advance, if so, you could just create a TextureAtlas and given the runtime conditions you pick a sprite from it. If the precomputation is doable at your end, then in this case you can still use sprites, and don't miss out on sprite batching.

2) With shaders you can definitely apply the effects you mention. To use them, you'll need to use a [MaterialMesh2dBundle](https://github.com/bevyengine/bevy/blob/a5d70b8952aed3616ffb5a58026ea6cb5375fc82/crates/bevy_sprite/src/mesh2d/material.rs#L537), instead of a SpriteBundle / SpriteSheetBundle. You can see an example [here](https://github.com/bevyengine/bevy/blob/a5d70b8952aed3616ffb5a58026ea6cb5375fc82/examples/shader/shader_material.rs), it's for 3d but in 2d is very similar.
In this case you can just have a Quad that has the dimension of your sprite, and in the shader sample one or more textures, apply gradients etc.

The downside is that you lose out on sprite batching. So it depends on how many quads with your custom material you are going to have on screen at any time. 
Just as a ballpark, if you are in the hundreds or less, I think you should be OK, but if you get in the thousands the performance may suffer, but of course YMMV.

There's maybe also the option to specialize the SpritePipeline, but I have not done it, and I think it's not possible out of the box, you'd need a fork, but others will know more on the subject.

Hope it was a bit helpful.
